File: src/types/resume.ts
export interface PersonalInfo {
  fullName: string;
  email: string;
  phone: string;
  location: string;
  linkedin: string;
  github: string;
  portfolio: string;
}

export interface Education {
  id: string;
  institution: string;
  degree: string;
  fieldOfStudy: string;
  startDate: string;
  endDate: string;
  gpa: string;
  location: string;
}

export interface WorkExperience {
  id: string;
  company: string;
  position: string;
  location: string;
  startDate: string;
  endDate: string;
  isCurrent: boolean;
  bullets: string[];
}

export interface Internship {
  id: string;
  company: string;
  role: string;
  location: string;
  startDate: string;
  endDate: string;
  isCurrent: boolean;
  bullets: string[];
}

export interface Project {
  id: string;
  title: string;
  description: string;
  technologies: string[];
  githubLink: string;
  liveLink: string;
  bullets: string[];
}

export interface SkillCategory {
  id: string;
  categoryName: string;
  skills: string[];
}

export interface Certification {
  id: string;
  name: string;
  issuer: string;
  issueDate: string;
  url: string;
}

export interface Achievement {
  id: string;
  title: string;
  description: string;
  date: string;
}

export interface Language {
  id: string;
  language: string;
  proficiency: string; // e.g., "Native", "Fluent", "Intermediate", "Basic"
}

export interface ResumeData {
  personalInfo: PersonalInfo;
  summary: string;
  education: Education[];
  experience: WorkExperience[];
  internships: Internship[];
  projects: Project[];
  skills: SkillCategory[];
  certifications: Certification[];
  achievements: Achievement[];
  languages: Language[];
}

export interface ResumeRecord {
  id: string;
  userId: string;
  title: string;
  templateId: string;
  data: ResumeData;
  atsScore: number;
  createdAt: string;
  updatedAt: string;
}

/**
 * Creates a clean, empty resume structure with zero hardcoded or fake data.
 */
export function createEmptyResumeData(): ResumeData {
  return {
    personalInfo: {
      fullName: "",
      email: "",
      phone: "",
      location: "",
      linkedin: "",
      github: "",
      portfolio: "",
    },
    summary: "",
    education: [],
    experience: [],
    internships: [],
    projects: [],
    skills: [],
    certifications: [],
    achievements: [],
    languages: [],
  };
}

/**
 * Normalizes partial or dynamically fetched/parsed resume data into a safe ResumeData structure.
 */
export function normalizeResumeData(data: Partial<ResumeData> | null | undefined): ResumeData {
  const empty = createEmptyResumeData();
  if (!data) return empty;

  return {
    personalInfo: {
      fullName: data.personalInfo?.fullName || "",
      email: data.personalInfo?.email || "",
      phone: data.personalInfo?.phone || "",
      location: data.personalInfo?.location || "",
      linkedin: data.personalInfo?.linkedin || "",
      github: data.personalInfo?.github || "",
      portfolio: data.personalInfo?.portfolio || "",
    },
    summary: data.summary || "",
    education: Array.isArray(data.education) ? data.education : [],
    experience: Array.isArray(data.experience) ? data.experience : [],
    internships: Array.isArray(data.internships) ? data.internships : [],
    projects: Array.isArray(data.projects) ? data.projects : [],
    skills: Array.isArray(data.skills) ? data.skills : [],
    certifications: Array.isArray(data.certifications) ? data.certifications : [],
    achievements: Array.isArray(data.achievements) ? data.achievements : [],
    languages: Array.isArray(data.languages) ? data.languages : [],
  };
}

File: backend/app/schemas/resume.py
from datetime import datetime
from typing import List, Optional
from pydantic import BaseModel, Field


class PersonalInfoSchema(BaseModel):
    fullName: str = ""
    email: str = ""
    phone: str = ""
    location: str = ""
    linkedin: str = ""
    github: str = ""
    portfolio: str = ""


class EducationSchema(BaseModel):
    id: str
    institution: str = ""
    degree: str = ""
    fieldOfStudy: str = ""
    startDate: str = ""
    endDate: str = ""
    gpa: str = ""
    location: str = ""


class WorkExperienceSchema(BaseModel):
    id: str
    company: str = ""
    position: str = ""
    location: str = ""
    startDate: str = ""
    endDate: str = ""
    isCurrent: bool = False
    bullets: List[str] = Field(default_factory=list)


class InternshipSchema(BaseModel):
    id: str
    company: str = ""
    role: str = ""
    location: str = ""
    startDate: str = ""
    endDate: str = ""
    isCurrent: bool = False
    bullets: List[str] = Field(default_factory=list)


class ProjectSchema(BaseModel):
    id: str
    title: str = ""
    description: str = ""
    technologies: List[str] = Field(default_factory=list)
    githubLink: str = ""
    liveLink: str = ""
    bullets: List[str] = Field(default_factory=list)


class SkillCategorySchema(BaseModel):
    id: str
    categoryName: str = ""
    skills: List[str] = Field(default_factory=list)


class CertificationSchema(BaseModel):
    id: str
    name: str = ""
    issuer: str = ""
    issueDate: str = ""
    url: str = ""


class AchievementSchema(BaseModel):
    id: str
    title: str = ""
    description: str = ""
    date: str = ""


class LanguageSchema(BaseModel):
    id: str
    language: str = ""
    proficiency: str = ""


class ResumeContentSchema(BaseModel):
    personalInfo: PersonalInfoSchema = Field(default_factory=PersonalInfoSchema)
    summary: str = ""
    education: List[EducationSchema] = Field(default_factory=list)
    experience: List[WorkExperienceSchema] = Field(default_factory=list)
    internships: List[InternshipSchema] = Field(default_factory=list)
    projects: List[ProjectSchema] = Field(default_factory=list)
    skills: List[SkillCategorySchema] = Field(default_factory=list)
    certifications: List[CertificationSchema] = Field(default_factory=list)
    achievements: List[AchievementSchema] = Field(default_factory=list)
    languages: List[LanguageSchema] = Field(default_factory=list)


class ResumeCreateSchema(BaseModel):
    title: str = "Untitled Resume"
    template_id: str = "modern"
    content: ResumeContentSchema = Field(default_factory=ResumeContentSchema)


class ResumeUpdateSchema(BaseModel):
    title: Optional[str] = None
    template_id: Optional[str] = None
    content: Optional[ResumeContentSchema] = None
    ats_score: Optional[int] = None


class ResumeResponseSchema(BaseModel):
    id: int
    user_id: int
    title: str
    template_id: str
    content: ResumeContentSchema
    ats_score: int
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


class AITailorRequestSchema(BaseModel):
    content: ResumeContentSchema
    target_role: str = "Software Engineer"


class AITailorResponseSchema(BaseModel):
    tailored_content: ResumeContentSchema
    ats_analysis: dict

